File: cuiserRice.rb
class CuiseurARiz
  def initialize
    @cuisson_en_cours = false
    @maintien_au_chaud = false
    @temps_maintien_au_chaud = 30  
    @surchauffe_temperature = 120  
  end

  def cuire_riz
    puts 'Cuisson en cours...'
    @cuisson_en_cours = true
    15.times do
      sleep(1)
      raise 'Surchauffe détectée. Arrêt de la cuisson.' if surchauffe_detectee
    end
    raise 'Le riz ne répond pas aux critères de qualité.' unless qualite_du_riz
  rescue StandardError => e
    puts "Erreur pendant la cuisson : #{e.message}"
  ensure
    @cuisson_en_cours = false
  end

  def maintien_chaud
    raise "Aucune cuisson en cours. Commencez d'abord la cuisson." unless @cuisson_en_cours

    puts "Le riz est maintenant maintenu au chaud pendant #{@temps_maintien_au_chaud} secondes."
    @maintien_au_chaud = true
    sleep(@temps_maintien_au_chaud)
    puts 'Le maintien au chaud est terminé.'
    @maintien_au_chaud = false
  rescue StandardError => e
    puts "Erreur pendant le maintien au chaud : #{e.message}"
  end

  def surchauffe_detectee
    temperature_actuelle = 130 
    raise 'Surchauffe détectée. Arrêt de la cuisson.' if temperature_actuelle >= @surchauffe_temperature

    false
  end

  def qualite_du_riz
    true
  end
end

cuiseur = CuiseurARiz.new

begin
  cuiseur.cuire_riz
rescue StandardError => e
  puts "Erreur générale pendant la cuisson : #{e.message}"
end

begin
  cuiseur.maintien_chaud
rescue StandardError => e
  puts "Erreur générale pendant le maintien au chaud : #{e.message}"
end

File: cuiserRice.ts
export class CuiseurARiz {
    public cuisson_en_cours: boolean = false;
    public maintien_au_chaud: boolean = false;
    public temps_maintien_au_chaud: number = 30; 
    public surchauffe_temperature: number = 120; 
  
    constructor() {}
  
    cuireRiz(): void {
        try {
            console.log("Cuisson en cours...");
            this.cuisson_en_cours = true;
      
            for (let i = 0; i < 15; i++) {
                this.surchauffeDetectee();
                console.log("Surchauffe non détectée. Continuation de la cuisson...");
          
            }
            if (!this.qualiteDuRiz()) {
                throw new Error("Le riz ne répond pas aux critères de qualité.");
            }
        } catch (error) {
            console.log(`Erreur pendant la cuisson : ${error.message}`);
        } finally {
            this.cuisson_en_cours = false;
        }
    }
  
    maintienChaud(): void {
        try {
            if (this.cuisson_en_cours) {
                console.log(`Le riz est maintenant maintenu au chaud pendant ${this.temps_maintien_au_chaud} secondes.`);
                this.maintien_au_chaud = true;
                console.log("Le maintien au chaud est terminé.");
                this.maintien_au_chaud = false;
            } else {
                throw new Error("Aucune cuisson en cours. Commencez d'abord la cuisson.");
            }
        } catch (error) {
            console.log(`Erreur pendant le maintien au chaud : ${error.message}`);
        }
    }
  
    public surchauffeDetectee(): void {
      
        const temperatureActuelle: number = 130; 
        if (temperatureActuelle >= this.surchauffe_temperature) {
            throw new Error("Surchauffe détectée. Arrêt de la cuisson.");
        }
    }
  
    public qualiteDuRiz(): boolean {
        return true;
    }
}
  
const cuiseur = new CuiseurARiz();
  
try {
    cuiseur.cuireRiz();
} catch (error) {
    console.log(`Erreur générale pendant la cuisson : ${error.message}`);
}
  
try {
    cuiseur.maintienChaud();
} catch (error) {
    console.log(`Erreur générale pendant le maintien au chaud : ${error.message}`);
}
  